_posts/2024-01-19-yannick-horrix-opentofu.md

```diff
@@ -0,0 +1,160 @@
+
+In this article we'll delve into a comparative analysis of OpenTofu and Terraform, exploring the impact of the license change, to help you make an informed decision when choosing the right tool for your infrastructure needs.
+# A Comparative Analysis: OpenTofu vs Terraform
+## Features
+In terms of functionality, both tools are currently equivalent. 
+The OpenTofu community has aimed to maintain feature parity with Terraform for the time being. 
+As such, both currently offer the same large set of features making them well suited tools for IaC workloads.
+
+Some small features were added with the release of OpenTofu 1.6.0 however, including:
+* An advanced testing feature for improved stability in configurations and modules
+* Enhanced S3 state backend with new authentication methods
+* A new provider and module registry, allowing a simpler method of publishing via pull request
+
+The upcoming version 1.7 aims to introduce more community requested features not available in Terraform, including:
+* Client-side state encryption for heightened security in regulated environments
+* Parameterizable backends, providers, and modules to enable more readable, DRY code
+* Third-party extensibility, with a plugin system for new state backends
+
+On Terraform's side, version 1.6 implemented performance improvements to the 'terraform test' command, as well as changes to the S3 backend.
+
+For now, both tools are functionally equivalent and compatible. 
+Time will tell whether this will remain the case as development continues on both sides.
+## Configuration and Syntax
+The core difference between OpenTofu and Terraform lies in its syntax. 
+OpenTofu uses a declarative syntax, where users can define their desired end state without having to worry about the steps required to get there. 
+Terraform, on the other hand, employs a procedural approach. 
+This means users have to define the specific steps, leading to higher control over the execution at the cost of higher complexity.
+
+In terms of configuration, OpenTofu focuses on simplicity in its configuration language, prioritizing readability and ease of use. 
+Terraform configuration, on the other hand, can be more verbose and complex to maintain.
+## Licensing and Pricing
+Both Terraform and OpenTofu are open-source tools, providing free usage.
+
+As said, OpenTofu uses the Mozilla Public License (MPL), which allows developers to freely use, modify, and distribute software for both commercial and non-commercial use.
+
+Terraform has now shifted to the Business Source License (BUSL). This also permits free use of source code, including commercial use, except when it’s used to provide an offering that competes with HashiCorp.
+An example of this could be running Terraform in a hosted way in CI/CD and offering this as a production service.
+
+So while, Terraform remains free to use in your projects, there can be an issue when incorporating Terraform in a service offering.
```

Comment:
This isn't even fully clarified nor disclosed by Hashicorp. The license opens end-users of Terraform up to potential license costs in the future, if Hashicorp wants to. 

The license is very restrictive and allows Hashicorp to "change the rules" along the way. You might be allowed to use Terraform today by Hashicorp, but there are no guarantees that that's the case tomorrow, even without a license change.

Reply:
E.g. using a Github Action for Terraform is very much a grey zone. Technically, only Hashicorp can create Github Actions, as creating one of your own would directly compete with Hashicorp as they [have an action available](https://github.com/marketplace/actions/hashicorp-setup-terraform).